logrotate.c

```diff
@@ -1221,6 +1226,28 @@ static int copyTruncate(const char *currLog, const char *saveLog, const struct s
                 goto fail;
             }
         }
+#if defined(WITH_FALLOCATE)
+    } else if (flags & LOG_FLAG_COPYREDUCE) {
+        message(MESS_DEBUG, "reducing %s\n", currLog);
+
+        if (!debug) {
+            if (fdsave >= 0)
+                fsync(fdsave);
+            block_bytes = ((off_t)read_bytes / sb->st_blksize) * sb->st_blksize;
+            if (block_bytes > 0) {
+                if (fallocate(fdcurr, FALLOC_FL_COLLAPSE_RANGE, 0L, block_bytes)) {
+                    message(MESS_ERROR, "error reducing %s: %s\n", currLog,
+                            strerror(errno));
+                    goto fail;
+                }
+            }
+            if (fallocate(fdcurr, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE, 0L, read_bytes - block_bytes)) {
+                message(MESS_ERROR, "error reducing %s: %s\n", currLog,
```

Comment:
Could you please make the diagnostic messages more specific, such that it is clear from the verbose output at which step the operation failed?

Reply:
got it.